### PR TITLE
VSTS Account Creation migration to AEX APIs

### PR DESCRIPTION
**Pull request by kuvinodms**

### CODE REVIEW

File: vsts_cd_manager/continuous_delivery_manager.py [hunk 1]
```diff
@@ -225,14 +223,14 @@ def _get_build_configuration(self, app_type_details):
...
+        # Find the identifier and set the properties.
+        # Default is externalGit
         type = 'Git'
         identifier = uri
         account_name = None
         team_project_name = None
-        auth_info = None
-        if username is not None and password is not None: 
-            auth_info = AuthorizationInfo('UsernamePassword', AuthorizationInfoParameters(None, None, username, password))
+        auth_info = AuthorizationInfo('UsernamePassword', AuthorizationInfoParameters(None, None, username, password))
```
Comment (bishal-pdMSFT) [inline vsts_cd_manager/continuous_delivery_manager.py:232]: Why this change?

File: vsts_cd_manager/continuous_delivery_manager.py [hunk 2]
```diff
@@ -113,30 +112,15 @@ def setup_continuous_delivery(self, swap_with_slot, app_type_details, cd_project
         account_url = 'https://{}.visualstudio.com'.format(quote(vsts_account_name))
         portalext_account_url = 'https://{}.portalext.visualstudio.com'.format(quote(vsts_account_name))
 
-        account_created = False
-        accountClient = Account('3.2-preview', None, self._azure_info.credentials)
+        # VSTS Account using AEX APIs
+        account_created = False
```
Comment (bishal-pdMSFT) [inline vsts_cd_manager/continuous_delivery_manager.py]: remove extra spaces. from here and elsewhere.

File: vsts_cd_manager/continuous_delivery_manager.py [hunk 3]
```diff
@@ -154,6 +138,20 @@ def setup_continuous_delivery(self, swap_with_slot, app_type_details, cd_project
...
+    def create_vsts_account(self, creds, vsts_account_name):
+        aex_url = 'https://app.vsaex.visualstudio.com'
+        accountClient = Account('4.0-preview.1', aex_url, creds)
+        self._update_progress(0, 100, 'Creating or getting Team Services account information')            
+        regions = accountClient.regions()
+        if regions.count == 0:
+            raise RuntimeError('Region details not found.')
+        region_name = regions.value[0].name
+        create_account_reponse = accountClient.create_account(vsts_account_name, region_name)
+        if create_account_reponse.id:
```
Comment (bishal-pdMSFT) [inline vsts_cd_manager/continuous_delivery_manager.py:150]: you are not checking whether account_exists or not any longer. is that not really helpful?

File: aex_accounts/account.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    def create_account(self, collection_name, preferred_region):
+
+        # Construct URL
+        url = '/_apis/hostacquisition/collections'
+
+        # Construct parameters
+        query_parameters = {}
+        query_parameters["api-version"] = self.api_version
+        query_parameters["collectionName"] = collection_name
+        query_parameters["preferredRegion"] = preferred_region
```
Comment (bishal-pdMSFT) [inline aex_accounts/account.py:36]: you should add headers like content-type

File: aex_accounts/account.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+        # Construct and send request
+        request = self._client.post(url, query_parameters)
+        response = self._client.send(request)
+
+        # Handle Response
+        deserialized = None
+        if response.status_code not in [200, 202]:
+            print("POST", request.url, file=stderr)
```
Comment (bishal-pdMSFT) [inline aex_accounts/account.py]: it is  a POST call but details are being sent in query parameter. is this okay?

File: aex_accounts/account.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            print("response:", response.status_code, file=stderr)
+            print(response.text, file=stderr)
+            raise HttpOperationError(self._deserialize, response)
+        else:
+            deserialized = self._deserialize('Collection', response)
```
Comment (bishal-pdMSFT) [inline aex_accounts/account.py:50]: check if deserialized is null and throw

File: aex_accounts/account.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    def create_account(self, collection_name, preferred_region):
+
+        # Construct URL
+        url = '/_apis/hostacquisition/collections'
+
+        # Construct parameters
+        query_parameters = {}
+        query_parameters["api-version"] = self.api_version
+        query_parameters["collectionName"] = collection_name
```
Comment (bishal-pdMSFT) [inline aex_accounts/account.py:35]: check for account name validity. it was being done in earlier code

File: aex_accounts/account.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        # Construct URL
+        url = '/_apis/hostacquisition/nameavailability/'+collection_name
+
+        # Construct and send request
+        request = self._client.get(url)
+        response = self._client.send(request)
+
+        # Handle Response
+        deserialized = None
+        if response.status_code not in [200, 202]:
```
Comment (bishal-pdMSFT) [inline aex_accounts/account.py]: 202 is not valid here

File: aex_accounts/account.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        # Construct URL
+        url = '/_apis/hostacquisition/regions'
+
+        # Construct and send request
+        request = self._client.get(url)
+        response = self._client.send(request)
+
+        # Handle Response
+        deserialized = None
+        if response.status_code not in [200, 202]:
```
Comment (bishal-pdMSFT) [inline aex_accounts/account.py]: 202 is not valid here

File: aex_accounts/account.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            print("POST", request.url, file=stderr)
+            print("response:", response.status_code, file=stderr)
+            print(response.text, file=stderr)
+            raise HttpOperationError(self._deserialize, response)
+        else:
+            deserialized = self._deserialize('Collection', response)
+
+        return deserialized
+
+    def is_account_available(self, collection_name):
```
Comment (bishal-pdMSFT) [inline aex_accounts/account.py]: where is this being used?

File: aex_accounts/version.py [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# coding=utf-8
+# --------------------------------------------------------------------------
+# Code generated by Microsoft (R) AutoRest Code Generator 1.0.1.0
+# Changes may cause incorrect behavior and will be lost if the code is
+# regenerated.
+# --------------------------------------------------------------------------
+
+VERSION = "4.0"
```
Comment (bishal-pdMSFT) [inline aex_accounts/version.py]: i don't understand how this version is related to Accounts. Should this be defined somewhere else?

File: aex_accounts/account.py [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        deserialized = None
+        if response.status_code not in [200, 202]:
+            print("POST", request.url, file=stderr)
+            print("response:", response.status_code, file=stderr)
+            print(response.text, file=stderr)
+            raise HttpOperationError(self._deserialize, response)
+        else:
+            deserialized = self._deserialize('Collection', response)
+
+        return deserialized
```
Comment (bishal-pdMSFT) [inline aex_accounts/account.py:52]: return object should be of type Account rather than collection

File: tests/test_continuous_delivery_manager.py [hunk 13]
```diff
@@ -68,35 +68,14 @@ def test_set_repository_info(self):
...
-        cdman.set_azure_web_info('group1', 'web1', 'fakeCreds', 'sub1', 'subname1', 'tenant1', 'South Central US')
-        cdman.set_repository_info('repoUrl1', 'master1', 'token1', None, None)
-        # call setup
-        with self.assertRaises(RuntimeError) as context:
-            cdman.setup_continuous_delivery('staging', 'AspNet', "https://account1.visualstudio.com", False, 'token2', None, None)
-        self.assertTrue('does not exist' in str(context.exception))
-
-    @patch("vsts_cd_manager.continuous_delivery_manager.ContinuousDelivery")
-    @patch("vsts_cd_manager.continuous_delivery_manager.Account")
     def test_setup_continuous_delivery___create_account(self, mock_account, mock_cd):
```
Comment (bishal-pdMSFT) [inline tests/test_continuous_delivery_manager.py:71]: do you want to add some negative test cases as well?

File: vsts_cd_manager/continuous_delivery_manager.py [hunk 14]
```diff
@@ -19,8 +19,7 @@
                                         CiArtifact, CiConfiguration, ProvisioningConfiguration,
                                         ProvisioningConfigurationSource, ProvisioningConfigurationTarget,
                                         SlotSwapConfiguration, SourceRepository, CreateOptions)
-from vsts_accounts import Account
-from vsts_accounts.models import (AccountCreateInfoInternal)
+from aex_accounts import Account
```
Comment (bishal-pdMSFT) [inline vsts_cd_manager/continuous_delivery_manager.py:22]: IMO, name the new class as  account. and rename old ones as legacy_account or something similar

**Review by bishal-pdMSFT — CHANGES_REQUESTED** [review]

File: aex_accounts/account.py (hunk 7 above)
Comment (kuvinodms) [reply aex_accounts/account.py:35]: create_collection apis anyways will throw error in case of invalid names. So, by not checking account validity we are saving 1 REST call on every request.

File: aex_accounts/account.py (hunk 4 above)
Comment (kuvinodms) [reply aex_accounts/account.py:36]: Since there is no Payload or Request-Body, adding content-type will be of no use.

File: aex_accounts/account.py (hunk 5 above)
Comment (kuvinodms) [reply aex_accounts/account.py]: Perfectly OK!
http://lists.w3.org/Archives/Public/ietf-http-wg/2010JulSep/0272.html

File: aex_accounts/account.py (hunk 6 above)
Comment (kuvinodms) [reply aex_accounts/account.py:50]: deserialized will never be NULL.
It will be None only in case of empty response data and in that case it will return None, which is expected.

File: aex_accounts/account.py (hunk 12 above)
Comment (kuvinodms) [reply aex_accounts/account.py:52]: There is no Model defined with the name Account. I've defined a model (aex_accounts/models/collection.py) with the name 'Collection' in the lines of expected response.

File: aex_accounts/account.py (hunk 10 above)
Comment (kuvinodms) [reply aex_accounts/account.py]: Nowhere. Since AEX is providing this API so I've coded it for future use.

File: aex_accounts/version.py (hunk 11 above)
Comment (kuvinodms) [reply aex_accounts/version.py]: This is AEX API version which is being used in setting up user-agent name.

File: vsts_cd_manager/continuous_delivery_manager.py (hunk 14 above)
Comment (kuvinodms) [reply vsts_cd_manager/continuous_delivery_manager.py:22]: I've name new class as Collection which is in consistence with what we've used in PE service and AEX both.

File: vsts_cd_manager/continuous_delivery_manager.py (hunk 3 above)
Comment (kuvinodms) [reply vsts_cd_manager/continuous_delivery_manager.py:150]: It will increase 1 extra REST call, anyways create_collection is doing it for us.
Also, we did the same in PE Service as well.

File: vsts_cd_manager/continuous_delivery_manager.py (hunk 1 above)
Comment (kuvinodms) [reply vsts_cd_manager/continuous_delivery_manager.py:232]: This is default auth_info for the case of External Git. No matter if username/password has some value or not, authorization_scheme will always be Username/Password. Verified the scenario with Ibiza as well.
Also, this particular change is already reviewed and pushed.
https://github.com/Microsoft/vsts-cd-manager/pull/18

File: tests/test_continuous_delivery_manager.py (hunk 13 above)
Comment (kuvinodms) [reply tests/test_continuous_delivery_manager.py:71]: Both positive and negative test scenarios are already in place.

File: vsts_cd_manager/continuous_delivery_manager.py (hunk 3 above)
Comment (bishal-pdMSFT) [reply vsts_cd_manager/continuous_delivery_manager.py:150]: if api is checking for existence and throwing proper error, then it is fine

File: vsts_cd_manager/continuous_delivery_manager.py (hunk 14 above)
Comment (bishal-pdMSFT) [reply vsts_cd_manager/continuous_delivery_manager.py:22]: Okay. But rename aex_accounts folder as just accounts and remove old files.

File: aex_accounts/version.py (hunk 11 above)
Comment (bishal-pdMSFT) [reply aex_accounts/version.py]: we decided to change it to azurecli/<pypi-package-version>

File: aex_accounts/account.py (hunk 10 above)
Comment (bishal-pdMSFT) [reply aex_accounts/account.py]: remove it

**Review by bishal-pdMSFT — APPROVED** [review]
